From 6608636dd3dfcacd28257b6ab49cf56ef435da19 Mon Sep 17 00:00:00 2001
From: Faisal Hassan <quic_faisalh@quicinc.com>
Date: Fri, 1 Nov 2024 20:24:53 +0530
Subject: [PATCH] phy: qcom: phy-qcom-snps-eusb2: Add support for SA8797P

This patch adds support for the SA8797P SoC to the phy-qcom-snps-eusb2
driver. The SA8797P platform allows multiple resources, such as clocks,
regulators, resets, TLMM, and interconnects, to be managed by firmware,
while the remaining responsibilities are handled by Linux.

The power domain, managed through the PM runtime framework, is utilized
for the activation and deactivation of these resources.

- Power domain "usb_core": Controls the GDSC, resets, and TLMM (GPIO).
- Power domain "usb_transfer": Controls clocks and interconnects.

In this variant, the two disjoint sets of resources are modeled against
the power domain suppliers, which can be turned on or off to achieve the
desired device states. Two modeled low power states, D1 and D3, are
supported.

transition: d0 <--> d3  :: both core and transfer suppliers
transition: d0 <--> d1  :: only transfer supplier

Patch set details :

Patch set 6 : Basic eUSB phy enablement, merged as .patch
Patch set 7 onwards : Add EUD check
Patch set 19  : merged .patch with repeater enabled

Change-Id: I1b7c0362fd42cd71ee6f7ad069ca5a92cce3270f
Signed-off-by: Faisal Hassan <quic_faisalh@quicinc.com>
---
 drivers/phy/qualcomm/phy-qcom-snps-eusb2.c | 245 +++++++++++++++++----
 1 file changed, 200 insertions(+), 45 deletions(-)

diff --git a/drivers/phy/qualcomm/phy-qcom-snps-eusb2.c b/drivers/phy/qualcomm/phy-qcom-snps-eusb2.c
index 1484691a41d5..2c6f91d7f3a2 100644
--- a/drivers/phy/qualcomm/phy-qcom-snps-eusb2.c
+++ b/drivers/phy/qualcomm/phy-qcom-snps-eusb2.c
@@ -10,6 +10,7 @@
 #include <linux/mod_devicetable.h>
 #include <linux/phy/phy.h>
 #include <linux/platform_device.h>
+#include <linux/pm_domain.h>
 #include <linux/regulator/consumer.h>
 #include <linux/reset.h>
 
@@ -115,6 +116,9 @@
 #define USB_PHY_APB_RDDATA_MSB		(0x148)
 #define APB_REG_RDDATA_15_8_MASK	GENMASK(7, 4)
 
+#define EUD_EN_REG			0x88E9000
+#define EUD_EN_SIZE			0x4
+
 static const char * const eusb2_hsphy_vreg_names[] = {
 	"vdd", "vdda12",
 };
@@ -125,6 +129,7 @@ struct qcom_snps_eusb2_hsphy {
 	struct phy *phy;
 	void __iomem *base;
 
+	struct device *dev;
 	struct clk *ref_clk;
 	struct reset_control *phy_reset;
 
@@ -133,8 +138,63 @@ struct qcom_snps_eusb2_hsphy {
 	enum phy_mode mode;
 
 	struct phy *repeater;
+
+	struct dev_pm_domain_list *pd_list;
+	bool fw_managed;
+	bool eud_enable;
 };
 
+static void qcom_snps_domain_detach(struct qcom_snps_eusb2_hsphy *phy)
+{
+	dev_pm_domain_detach_list(phy->pd_list);
+}
+
+static int qcom_snps_domain_attach(struct qcom_snps_eusb2_hsphy *phy)
+{
+	struct dev_pm_domain_attach_data pd_data = {
+		.pd_flags       = PD_FLAG_NO_DEV_LINK,
+		.pd_names       = (const char*[]) { "transfer", "core" },
+		.num_pd_names   = 2,
+	};
+	struct device *dev = phy->dev;
+	int ret;
+
+	ret = dev_pm_domain_attach_list(dev, &pd_data, &phy->pd_list);
+	if (ret < 0) {
+		dev_err(dev, "domain attach failed %d)\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+/* d3_to_d0 transition by turning on all the suppliers */
+static int qcom_snps_d3_to_d0(struct qcom_snps_eusb2_hsphy *phy)
+{
+	int ret;
+
+	if (!phy->pd_list || phy->eud_enable)
+		return 0;
+
+	ret = pm_runtime_resume_and_get(phy->pd_list->pd_devs[1]);
+	if (ret)
+		return ret;
+
+	ret = pm_runtime_resume_and_get(phy->pd_list->pd_devs[0]);
+
+	return ret;
+}
+
+/* d0_to_d3 transition by turning off all the suppliers */
+static void qcom_snps_d0_to_d3(struct qcom_snps_eusb2_hsphy *phy)
+{
+	if (!phy->pd_list || phy->eud_enable)
+		return;
+
+	pm_runtime_put_sync(phy->pd_list->pd_devs[0]);
+	pm_runtime_put_sync(phy->pd_list->pd_devs[1]);
+}
+
 static int qcom_snps_eusb2_hsphy_set_mode(struct phy *p, enum phy_mode mode, int submode)
 {
 	struct qcom_snps_eusb2_hsphy *phy = phy_get_drvdata(p);
@@ -235,36 +295,53 @@ static int qcom_snps_eusb2_hsphy_init(struct phy *p)
 	struct qcom_snps_eusb2_hsphy *phy = phy_get_drvdata(p);
 	int ret;
 
-	ret = regulator_bulk_enable(ARRAY_SIZE(phy->vregs), phy->vregs);
-	if (ret)
-		return ret;
+	if(phy->eud_enable)
+		return 0;
 
-	ret = phy_init(phy->repeater);
-	if (ret) {
-		dev_err(&p->dev, "repeater init failed. %d\n", ret);
-		goto disable_vreg;
-	}
-
-	ret = clk_prepare_enable(phy->ref_clk);
-	if (ret) {
-		dev_err(&p->dev, "failed to enable ref clock, %d\n", ret);
-		goto disable_vreg;
-	}
-
-	ret = reset_control_assert(phy->phy_reset);
-	if (ret) {
-		dev_err(&p->dev, "failed to assert phy_reset, %d\n", ret);
-		goto disable_ref_clk;
+	ret = qcom_snps_d3_to_d0(phy);
+	if (ret < 0) {
+		dev_err(&p->dev, "Failed to transition to d0 state\n");
+		return ret;
 	}
 
-	usleep_range(100, 150);
-
-	ret = reset_control_deassert(phy->phy_reset);
-	if (ret) {
-		dev_err(&p->dev, "failed to de-assert phy_reset, %d\n", ret);
-		goto disable_ref_clk;
+	if (!phy->fw_managed) {
+		ret = regulator_bulk_enable(ARRAY_SIZE(phy->vregs), phy->vregs);
+		if (ret)
+			return ret;
+
+		ret = phy_init(phy->repeater);
+		if (ret) {
+			dev_err(&p->dev, "repeater init failed. %d\n", ret);
+			goto disable_vreg;
+		}
+
+		ret = clk_prepare_enable(phy->ref_clk);
+		if (ret) {
+			dev_err(&p->dev, "failed to enable ref clock, %d\n", ret);
+			goto disable_vreg;
+		}
+
+		ret = reset_control_assert(phy->phy_reset);
+		if (ret) {
+			dev_err(&p->dev, "failed to assert phy_reset, %d\n", ret);
+			goto disable_ref_clk;
+		}
+
+		usleep_range(100, 150);
+
+		ret = reset_control_deassert(phy->phy_reset);
+		if (ret) {
+			dev_err(&p->dev, "failed to de-assert phy_reset, %d\n", ret);
+			goto disable_ref_clk;
+		}
+	} else {
+		ret = phy_init(phy->repeater);
+		if (ret) {
+			dev_err(&p->dev, "repeater init failed. %d\n", ret);
+			qcom_snps_d0_to_d3(phy);
+			return ret;
+		}
 	}
-
 	qcom_snps_eusb2_hsphy_write_mask(phy->base, USB_PHY_CFG0,
 					 CMN_CTRL_OVERRIDE_EN, CMN_CTRL_OVERRIDE_EN);
 
@@ -281,10 +358,30 @@ static int qcom_snps_eusb2_hsphy_init(struct phy *p)
 	qcom_snps_eusb2_hsphy_write_mask(phy->base, USB_PHY_FSEL_SEL,
 					 FSEL_SEL, FSEL_SEL);
 
-	/* update ref_clk related registers */
-	ret = qcom_eusb2_ref_clk_init(phy);
-	if (ret)
-		goto disable_ref_clk;
+	if (!phy->fw_managed) {
+		/* update ref_clk related registers */
+		ret = qcom_eusb2_ref_clk_init(phy);
+		if (ret)
+			goto disable_ref_clk;
+	}
+
+	/* Program below registers assuming highest freq  */
+	qcom_snps_eusb2_hsphy_write_mask(phy->base, USB_PHY_HS_PHY_CTRL_COMMON0,
+					 FSEL_MASK,
+					 FIELD_PREP(FSEL_MASK, FSEL_38_4_MHZ_VAL));
+
+	qcom_snps_eusb2_hsphy_write_mask(phy->base, USB_PHY_CFG_CTRL_2,
+					 PHY_CFG_PLL_FB_DIV_7_0_MASK,
+					 DIV_7_0_38_4_MHZ_VAL);
+
+	qcom_snps_eusb2_hsphy_write_mask(phy->base, USB_PHY_CFG_CTRL_3,
+					 PHY_CFG_PLL_FB_DIV_11_8_MASK,
+					 DIV_11_8_38_4_MHZ_VAL);
+
+	qcom_snps_eusb2_hsphy_write_mask(phy->base, USB_PHY_CFG_CTRL_3,
+					 PHY_CFG_PLL_REF_DIV, PLL_REF_DIV_VAL);
+
+	/* Program below registers assuming highest freq */
 
 	qcom_snps_eusb2_hsphy_write_mask(phy->base, USB_PHY_CFG_CTRL_1,
 					 PHY_CFG_PLL_CPBIAS_CNTRL_MASK,
@@ -348,12 +445,15 @@ static int qcom_snps_eusb2_hsphy_exit(struct phy *p)
 {
 	struct qcom_snps_eusb2_hsphy *phy = phy_get_drvdata(p);
 
-	clk_disable_unprepare(phy->ref_clk);
+	qcom_snps_d0_to_d3(phy);
 
-	regulator_bulk_disable(ARRAY_SIZE(phy->vregs), phy->vregs);
+	if (!phy->fw_managed) {
+		clk_disable_unprepare(phy->ref_clk);
 
-	phy_exit(phy->repeater);
+		regulator_bulk_disable(ARRAY_SIZE(phy->vregs), phy->vregs);
 
+	}
+	phy_exit(phy->repeater);
 	return 0;
 }
 
@@ -364,6 +464,24 @@ static const struct phy_ops qcom_snps_eusb2_hsphy_ops = {
 	.owner		= THIS_MODULE,
 };
 
+static int qcom_snps_eusb2_enable_eud(struct qcom_snps_eusb2_hsphy *phy)
+{
+	void __iomem *eud_en;
+	u32 val;
+
+	eud_en = ioremap(EUD_EN_REG, EUD_EN_SIZE);
+	if (!eud_en)
+		return -ENOMEM;
+
+	val = readl_relaxed(eud_en);
+	if(val)
+		phy->eud_enable = true;
+
+	iounmap(eud_en);
+
+	return 0;
+}
+
 static int qcom_snps_eusb2_hsphy_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
@@ -378,32 +496,57 @@ static int qcom_snps_eusb2_hsphy_probe(struct platform_device *pdev)
 	if (!phy)
 		return -ENOMEM;
 
+	if (of_device_is_compatible(dev->of_node,
+				    "qcom,sa8797p-snps-eusb2-phy-fw-managed"))
+		phy->fw_managed = true;
+
 	phy->base = devm_platform_ioremap_resource(pdev, 0);
 	if (IS_ERR(phy->base))
 		return PTR_ERR(phy->base);
 
-	phy->phy_reset = devm_reset_control_get_exclusive(dev, NULL);
-	if (IS_ERR(phy->phy_reset))
-		return PTR_ERR(phy->phy_reset);
+	phy->dev = dev;
+
+	if (phy->fw_managed) {
+		ret = qcom_snps_domain_attach(phy);
+		if (ret)
+			return ret;
+	} else {
+		phy->phy_reset = devm_reset_control_get_exclusive(dev, NULL);
+		if (IS_ERR(phy->phy_reset))
+			return PTR_ERR(phy->phy_reset);
+
+		phy->ref_clk = devm_clk_get(dev, "ref");
+		if (IS_ERR(phy->ref_clk))
+			return dev_err_probe(dev, PTR_ERR(phy->ref_clk),
+					     "failed to get ref clk\n");
+	}
 
-	phy->ref_clk = devm_clk_get(dev, "ref");
-	if (IS_ERR(phy->ref_clk))
-		return dev_err_probe(dev, PTR_ERR(phy->ref_clk),
-				     "failed to get ref clk\n");
+	if (of_property_read_bool(np,"qcom,enable-eud")) {
+		ret = qcom_snps_eusb2_enable_eud(phy);
+		if (ret) {
+			dev_info(dev, "failed to map eud reg");
+			return ret;
+		}
+	}
 
 	num = ARRAY_SIZE(phy->vregs);
 	for (i = 0; i < num; i++)
 		phy->vregs[i].supply = eusb2_hsphy_vreg_names[i];
 
-	ret = devm_regulator_bulk_get(dev, num, phy->vregs);
-	if (ret)
-		return dev_err_probe(dev, ret,
-				     "failed to get regulator supplies\n");
+
+	if (!phy->fw_managed) {
+		ret = devm_regulator_bulk_get(dev, num, phy->vregs);
+		if (ret)
+			return dev_err_probe(dev, ret,
+					     "failed to get regulator supplies\n");
+	}
 
 	phy->repeater = devm_of_phy_get_by_index(dev, np, 0);
-	if (IS_ERR(phy->repeater))
+	if (IS_ERR(phy->repeater)) {
+		qcom_snps_domain_detach(phy);
 		return dev_err_probe(dev, PTR_ERR(phy->repeater),
 				     "failed to get repeater\n");
+	}
 
 	generic_phy = devm_phy_create(dev, NULL, &qcom_snps_eusb2_hsphy_ops);
 	if (IS_ERR(generic_phy)) {
@@ -423,14 +566,26 @@ static int qcom_snps_eusb2_hsphy_probe(struct platform_device *pdev)
 	return 0;
 }
 
+static void qcom_snps_eusb2_hsphy_remove(struct platform_device *pdev)
+{
+	struct qcom_snps_eusb2_hsphy *phy = platform_get_drvdata(pdev);
+
+	if (phy->fw_managed) {
+		qcom_snps_d0_to_d3(phy);
+		qcom_snps_domain_detach(phy);
+	}
+}
+
 static const struct of_device_id qcom_snps_eusb2_hsphy_of_match_table[] = {
 	{ .compatible = "qcom,sm8550-snps-eusb2-phy", },
+	{ .compatible = "qcom,sa8797p-snps-eusb2-phy-fw-managed", },
 	{ },
 };
 MODULE_DEVICE_TABLE(of, qcom_snps_eusb2_hsphy_of_match_table);
 
 static struct platform_driver qcom_snps_eusb2_hsphy_driver = {
 	.probe		= qcom_snps_eusb2_hsphy_probe,
+	.remove_new	= qcom_snps_eusb2_hsphy_remove,
 	.driver = {
 		.name	= "qcom-snps-eusb2-hsphy",
 		.of_match_table = qcom_snps_eusb2_hsphy_of_match_table,
-- 
2.34.1

